### ISSUE TITLE
Http1xClientConnection holds unnecessary references to ended client request/response objects

### ISSUE THREAD

**Issue by michaelpog** [vert-x3/issues#375]
When an HTTP connection (Holder<C>) is recycled and put back into the Pool<C>, that connection still holds, indirectly a reference to the HttpClientRequestImpl object, and the entire context associated with that request. (ResponseHandler etc).
![screen shot 2018-05-24 at 10 58 08 am](https://user-images.githubusercontent.com/7913776/40519305-bdb66c80-5f73-11e8-82ea-c639d4ff1758.png)
In the below screenshot you can see that a connection that is acquired from the pool, still contains all the old request + response data.
![screen shot 2018-05-24 at 11 04 21 am](https://user-images.githubusercontent.com/7913776/40519344-e3abe47e-5f73-11e8-8ccb-39a7d6182180.png)

The issue with that is that connection may stay in the pool for up to the "idle time" specified by the user, in the HttpClientOptions. 
In our case that idle time is 60 seconds.
With our current allocation rate, we have a young collection every 20-30 seconds.
That means that a very large number of objects that should be simply reclaimed by the GC, during the young collection, actually get promoted and copied to the survivor space.
That completely kills our performs as it increases our GC pauses due to copying of those object to almost 100ms.

Looking closer
![screen shot 2018-05-24 at 3 06 24 pm](https://user-images.githubusercontent.com/7913776/40519418-3c741cac-5f74-11e8-8ca5-97d7181aa140.png)
It seems like there is a circular dependency Holder - > Http1xClientConnetion -> ConnectionListener -> Holder.
And the ConnectionListener should actually be reset (set to null) when the Holder is put back to the pool, and recreated when the connection is reacquired.

I started working on a pull request https://github.com/eclipse/vert.x/pull/2475 ( I currently have some tests broken). This code is pretty sensitive, I rely on @vietj to decide what to do next.

**Comment by vietj** [comment]
hi,

when the connection is evicted from the pool, the memory can be reclaimed.

so this is not a leak per se.

**Comment by vietj** [comment]
The circular references you are describing : Holder - > Http1xClientConnetion -> ConnectionListener -> Holder are fine, since the holder reference itself and not another holder

**Comment by vietj** [comment]
it seems that what you are observing is similar to a lambda capture that would occur in Pool#createConnection method when the connection listener is created.

The listener references the waiter passed to the method. With my JDK I don't have this:

<img width="850" alt="screen shot 2018-05-25 at 08 35 34" src="https://user-images.githubusercontent.com/225674/40529792-a237a978-5ff6-11e8-87a0-b6fc8a1a97a8.png">

what is your actual JDK version ?

**Comment by vietj** [comment]
actually I do have it as well on 3.5 but it's not in master.

**Comment by vietj** [comment]
I will port this commit to 3.5 from master that removes the dependency of the ConnectionListener onto the waiter : https://github.com/eclipse/vert.x/commit/f894f4433c6aa86257412edd7af58ae2e5597b79#diff-f9568f2488b0f7e86b6a73c4120acf86

in short it removes from the ConnectionListener interface the connect callback and only this callback needs to reference the waiter.

**Comment by vietj** [comment]
see https://github.com/eclipse/vert.x/issues/2477

**Comment by michaelpog** [comment]
Thanks @vietj , looks like I used the wrong terminology, and you're right. It's not a memory leak, but it does keep the objects for longer than it should be, which puts pressure on GC, which is what I really meant to say. Thanks for the correction.
The pr looks good. I don't see past that line of code https://github.com/eclipse/vert.x/commit/f894f4433c6aa86257412edd7af58ae2e5597b79#diff-3ffe38f201de7b0ab11f0d467a78c7d2R255 so as long as that clientHandler doesn't get retained for longer than needed (and keeps that future object with the reference to the waiter), then I think it will work. And I was definitely not on the right track.

Please let me know when it's merged to a release and what version I should use to test it.

**Comment by vietj** [comment]
@michaelpog it has been merged in 3.5.x branch so you can try it, it shall be released within a couple of days in a CR3

**Comment by michaelpog** [comment]
Is it in the 3.5.2-SNAPSHOT already? Or 3.5.2.CR2? Or just 3.5.1/3.5.0?

**Comment by vietj** [comment]
it is in 3.5.2-SNAPSHOT

**Comment by michaelpog** [comment]
@vietj an update. I've tested and that issue looks to be fixed. But unfortunately that did not fix the high GC pauses. I've gone back to metrics we had from the time I deployed our application with 3.5.0 and I did not see the same GC pause issues. I'm investigating further. I'll keep you posted. The heap dump does not show any "leaks" at the moment, which is good news.

The symptom is still the same, a lot of objects get promoted to survivor spaces
